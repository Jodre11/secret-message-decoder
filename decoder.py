import requests
from bs4 import BeautifulSoup
from typing import Dict, Tuple
from dataclasses import dataclass
from requests.exceptions import RequestException

@dataclass
class Coordinate:
    """Represents a single coordinate in the ASCII art grid."""
    x: int
    y: int
    char: str

class Grid:
    """Represents the ASCII art grid with its dimensions and characters."""
    def __init__(self, coordinates: Dict[Tuple[int, int], str], max_x: int, max_y: int):
        self.coordinates = coordinates
        self.max_x = max_x
        self.max_y = max_y
    
    def get_char(self, x: int, y: int) -> str:
        """Get the character at the specified coordinates, or space if none exists."""
        return self.coordinates.get((x, y), ' ')
    
    def to_string(self) -> str:
        """Convert the grid to a string representation."""
        rows = []
        for y in range(self.max_y, -1, -1):
            row = ''.join(self.get_char(x, y) for x in range(self.max_x + 1))
            # Only strip trailing spaces from the first and last rows
            if y == self.max_y or y == 0:
                row = row.rstrip()
            rows.append(row)
        return '\n'.join(rows)

def parse_coordinates(doc_url: str) -> Grid:
    """
    Parse coordinates and characters from a Google Doc URL.
    
    Args:
        doc_url (str): URL of the Google Doc containing the character coordinates
        
    Returns:
        Grid: A Grid object containing the parsed coordinates and dimensions
        
    Raises:
        ValueError: If no table is found in the document
        RequestException: If there's an error fetching the document
        
    Example:
        >>> grid = parse_coordinates("https://example.com/doc")
    """
    try:
        # Fetch the document content
        response = requests.get(doc_url)
        response.raise_for_status()  # Raise exception for bad status codes
    except RequestException as e:
        raise RequestException(f"Failed to fetch document: {e}") from e
        
    soup = BeautifulSoup(response.text, 'html.parser')
    
    # Find the table in the document
    table = soup.find('table')
    if not table:
        raise ValueError("No table found in the document")
    
    coordinates: Dict[Tuple[int, int], str] = {}
    max_x = 0
    max_y = 0
    
    # Skip the header row
    rows = table.find_all('tr')[1:]
    for row in rows:
        cols = row.find_all('td')
        if len(cols) != 3:
            continue
            
        try:
            x = int(cols[0].text.strip())
            char = cols[1].text.strip()
            y = int(cols[2].text.strip())
        except (ValueError, IndexError) as e:
            raise ValueError(f"Invalid coordinate data in row: {e}") from e
        
        coordinates[(x, y)] = char
        max_x = max(max_x, x)
        max_y = max(max_y, y)
    
    return Grid(coordinates, max_x, max_y)

def decode_secret_message(doc_url: str) -> None:
    """
    Decodes a secret message from a Google Doc containing character coordinates.
    
    Args:
        doc_url (str): URL of the Google Doc containing the character coordinates
        
    Example:
        >>> decode_secret_message("https://example.com/doc")
        █▀▀▀
        █▀▀ 
        █
    """
    grid = parse_coordinates(doc_url)
    print(grid.to_string())

if __name__ == "__main__":
    # Example usage
    example_url = "https://docs.google.com/document/d/e/2PACX-1vQGUck9HIFCyezsrBSnmENk5ieJuYwpt7YHYEzeNJkIb9OSDdx-ov2nRNReKQyey-cwJOoEKUhLmN9z/pub"
    decode_secret_message(example_url) 