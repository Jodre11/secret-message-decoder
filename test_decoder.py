import pytest
from decoder import decode_secret_message, parse_coordinates, Grid
from requests.exceptions import RequestException

class MockResponse:
    def __init__(self, text, status_code=200):
        self.text = text
        self.status_code = status_code
    
    def raise_for_status(self):
        if self.status_code >= 400:
            raise RequestException(f"HTTP {self.status_code}")

def test_parse_coordinates():
    # Mock HTML content that represents the letter 'F'
    mock_html = """
    <table>
        <tr><td>x-coordinate</td><td>Character</td><td>y-coordinate</td></tr>
        <tr><td>0</td><td>█</td><td>0</td></tr>
        <tr><td>0</td><td>█</td><td>1</td></tr>
        <tr><td>0</td><td>█</td><td>2</td></tr>
        <tr><td>1</td><td>▀</td><td>1</td></tr>
        <tr><td>1</td><td>▀</td><td>2</td></tr>
        <tr><td>2</td><td>▀</td><td>1</td></tr>
        <tr><td>2</td><td>▀</td><td>2</td></tr>
        <tr><td>3</td><td>▀</td><td>2</td></tr>
    </table>
    """
    
    # Mock the requests.get function
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse(mock_html)
    
    try:
        # Call the function
        grid = parse_coordinates("https://example.com")
        
        # Verify the coordinates
        assert grid.coordinates == {
            (0, 0): '█',
            (0, 1): '█',
            (0, 2): '█',
            (1, 1): '▀',
            (1, 2): '▀',
            (2, 1): '▀',
            (2, 2): '▀',
            (3, 2): '▀'
        }
        
        # Verify max coordinates
        assert grid.max_x == 3
        assert grid.max_y == 2
        
        # Verify string representation
        assert grid.to_string() == "█▀▀▀\n█▀▀ \n█"
    finally:
        # Restore the original requests.get
        requests.get = original_get

def test_parse_coordinates_empty_document():
    # Mock empty HTML content
    mock_html = "<html><body></body></html>"
    
    # Mock the requests.get function
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse(mock_html)
    
    try:
        # Verify that the function raises ValueError for empty document
        with pytest.raises(ValueError, match="No table found in the document"):
            parse_coordinates("https://example.com")
    finally:
        # Restore the original requests.get
        requests.get = original_get

def test_parse_coordinates_http_error():
    # Mock the requests.get function to raise an error
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse("", status_code=404)
    
    try:
        # Verify that the function raises RequestException for HTTP error
        with pytest.raises(RequestException):
            parse_coordinates("https://example.com")
    finally:
        # Restore the original requests.get
        requests.get = original_get

def test_parse_coordinates_invalid_data():
    # Mock HTML content with invalid coordinate data
    mock_html = """
    <table>
        <tr><td>x-coordinate</td><td>Character</td><td>y-coordinate</td></tr>
        <tr><td>not a number</td><td>█</td><td>0</td></tr>
    </table>
    """
    
    # Mock the requests.get function
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse(mock_html)
    
    try:
        # Verify that the function raises ValueError for invalid data
        with pytest.raises(ValueError, match="Invalid coordinate data"):
            parse_coordinates("https://example.com")
    finally:
        # Restore the original requests.get
        requests.get = original_get

def test_decode_secret_message(capsys, snapshot):
    # Mock URL for testing
    test_url = "https://docs.google.com/document/d/test"
    
    # Mock HTML content that represents the letter 'F'
    mock_html = """
    <table>
        <tr><td>x-coordinate</td><td>Character</td><td>y-coordinate</td></tr>
        <tr><td>0</td><td>█</td><td>0</td></tr>
        <tr><td>0</td><td>█</td><td>1</td></tr>
        <tr><td>0</td><td>█</td><td>2</td></tr>
        <tr><td>1</td><td>▀</td><td>1</td></tr>
        <tr><td>1</td><td>▀</td><td>2</td></tr>
        <tr><td>2</td><td>▀</td><td>1</td></tr>
        <tr><td>2</td><td>▀</td><td>2</td></tr>
        <tr><td>3</td><td>▀</td><td>2</td></tr>
    </table>
    """
    
    # Mock the requests.get function
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse(mock_html)
    
    try:
        # Call the function
        decode_secret_message(test_url)
        
        # Capture the output
        captured = capsys.readouterr()
        output = captured.out.strip()
        
        # Verify against snapshot
        assert output == snapshot
    finally:
        # Restore the original requests.get
        requests.get = original_get

def test_empty_document():
    # Mock URL for testing
    test_url = "https://docs.google.com/document/d/test"
    
    # Mock empty HTML content
    mock_html = "<html><body></body></html>"
    
    # Mock the requests.get function
    import requests
    original_get = requests.get
    requests.get = lambda url: MockResponse(mock_html)
    
    try:
        # Verify that the function raises ValueError for empty document
        with pytest.raises(ValueError, match="No table found in the document"):
            decode_secret_message(test_url)
    finally:
        # Restore the original requests.get
        requests.get = original_get 