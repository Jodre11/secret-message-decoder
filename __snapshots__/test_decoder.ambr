# serializer version: 1
# name: test_decode_secret_message
  '''
  █▀▀▀
  █▀▀ 
  █
  '''
# ---
